noir/noir-repo/acvm-repo/brillig_vm/src/memory.rs

```diff
@@ -143,10 +234,77 @@ impl TryFrom<MemoryValue> for bool {
 
     fn try_from(memory_value: MemoryValue) -> Result<Self, Self::Error> {
         memory_value.expect_bit_size(1)?;
+        let as_integer = memory_value.extract_integer().unwrap();
+
+        if as_integer.is_zero() {
+            Ok(false)
+        } else if as_integer.is_one() {
+            Ok(true)
+        } else {
+            unreachable!("value typed as bool is greater than one")
+        }
+    }
+}
+
+impl TryFrom<&MemoryValue> for FieldElement {
+    type Error = MemoryTypeError;
+
+    fn try_from(memory_value: &MemoryValue) -> Result<Self, Self::Error> {
+        memory_value.expect_bit_size(FieldElement::max_num_bits())?;
+        Ok(memory_value.extract_field().cloned().expect("Field memory cell is not a field"))
+    }
+}
+
+impl TryFrom<&MemoryValue> for u64 {
+    type Error = MemoryTypeError;
+
+    fn try_from(memory_value: &MemoryValue) -> Result<Self, Self::Error> {
+        memory_value.expect_bit_size(64)?;
+        Ok(memory_value
+            .extract_integer()
+            .expect("Integer memory cell is not an integer")
+            .try_into()
+            .unwrap())
+    }
```

Comment:
We can avoid having two calls to extract the bit size and the value by using `let ... else` here.

We should also have an error for reading field memory as if it were an integer.

Reply:
Hmmm I think we cannot use num-traits to generalize this since they don't have a bit size trait that we need to validate the bit size... maybe we could add a trait with the bit size but at that point I think it's not worth it

Reply:
Refactored it to make it a one liner and remove the extra expect, but I still need to implement it for every numeric type hmmm

Reply:
> no bit size method.

Huh, that's very odd. Agreed on it not being worth it at that point for the number of copies we'd have.

Reply:
I guess there's always `std::mem::size_of` but nbd.

Reply:
Actually that wouldn't work for u1 anyway.